velox/functions/prestosql/Sequence.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/ConstantVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+template <typename TInput1, typename TInput2>
+class SequenceFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10'000;
+
+ private:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    try {
+      if (args[0]->isConstantEncoding() && args[1]->isConstantEncoding()) {
+        localResult = applyConstant(rows, args, outputType, context);
+      } else {
+        localResult = applyVector(rows, args, outputType, context);
+      }
+      context.moveOrCopyResult(localResult, rows, result);
+    } catch (const std::exception& e) {
+      context.setErrors(rows, std::current_exception());
+      return;
+    }
+  }
+
+  VectorPtr applyConstant(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    auto startVector = args[0]->as<ConstantVector<int64_t>>();
+    auto endVector = args[1]->as<ConstantVector<int64_t>>();
+    if (startVector->isNullAt(0) || endVector->isNullAt(0)) {
+      // If start or end is a null constant, the result should be all nulls.
+      return BaseVector::createNullConstant(outputType, numRows, pool);
+    }
+
+    int start = startVector->valueAt(0);
+    int end = endVector->valueAt(0);
+    const auto elementsCount = end - start + 1;
+    if (elementsCount <= 0) {
+      return BaseVector::createNullConstant(outputType, numRows, pool);
```

Comment:
Is this the right semantic? Here is what I see in Presto:

```
presto> select sequence(10, 5);
        _col0
---------------------
 [10, 9, 8, 7, 6, 5]

presto> select sequence(5, 10);
        _col0
---------------------
 [5, 6, 7, 8, 9, 10]

```

Reply:
Good catch!